### PR TITLE
MAINT: Use `.identifier = val` to fill type structs

### PR DESCRIPTION
**Pull request by seberg**
I find the `.tp_name` identifier/designator method *much* easier to read. Python only uses in one or two places where it is explicitly a `static`. So before I go ahead and change it everywhere, are all of these usages: "static storage duration" or is there some reason that the way we create them is considered "dynamic storage duration", in which case initialization would not be defined by the C99 standard.

### CODE REVIEW

**Comment by seberg** [comment]
OK, its complete for the type objects now. The number methods, etc. are another worthy target (and maybe even more obvious).

**Comment by seberg** [comment]
Seemed to me like it was just not valid C89, so went ahead and changed a lot more (rationaltests.c.src is probably the only file with serious chunks of old notation)

File: numpy/core/src/multiarray/number.c [hunk 1]
```diff
@@ -923,47 +923,43 @@ array_index(PyArrayObject *v)
...
+    .nb_inplace_xor = (binaryfunc)array_inplace_bitwise_xor,
+    .nb_inplace_or = (binaryfunc)array_inplace_bitwise_or,
+
+    .nb_floor_divide = (binaryfunc)array_floor_divide,
+    .nb_true_divide = (binaryfunc)array_true_divide,
+    .nb_inplace_floor_divide = (binaryfunc)array_inplace_floor_divide,
+    .nb_inplace_true_divide = (binaryfunc)array_inplace_true_divide,
+
+    .nb_matrix_multiply = (binaryfunc)array_matrix_multiply,
+    .nb_inplace_matrix_multiply = (binaryfunc)array_inplace_matrix_multiply,
```
Comment (seberg) [inline numpy/core/src/multiarray/number.c]: This on struct is slightly reordered, (`.nb_index` together with int and float), may make sense for others, but as a follow-u[

**Comment by eric-wieser** [comment]
Perhaps worth including "C99 designated initializers" in the commit message

**Review by eric-wieser — APPROVED** [review]
Haven't checked the struct members rigorously, but I like the idea and trust you got it right, and CI would catch it had you not.

Does CPython do this anywhere internally? If not, it might be worth checking their mailing list archive for a reason they didn't.

Edit: [yes, they do](https://mail.python.org/pipermail/python-dev/2017-January/147154.html) (see [PEP7](https://www.python.org/dev/peps/pep-0007/#c-dialect))

**Comment by seberg** [comment]
Ah, did not stumble over that discussion, should have googled more python specific. I reworded this, and it reads like C-Python was just a bit hesitant due to code churn. Errors here are very unlikely to not be crashes.

@eric-wieser I assume you are OK if I self merge after giving it another pass myself pass later?

**Comment by mattip** [comment]
Thanks @seberg, this is a nice cleanup

**Pull request merged**